DietCake Board 1.0

rdaitan opened the pull request.

maragatchalian commented on app/controllers/comment_controller.php:
```diff
@@ -0,0 +1,33 @@
+
+        switch ($page) {
+            case 'create_end':
+                $comment            = new Comment;
```
`$comment = new Comment();`
Use this format even without passing any arguments. Please apply to all po :bow:

rdaitan replied:
Done po.

maragatchalian commented on app/controllers/thread_controller.php:
```diff
@@ -0,0 +1,78 @@
+    public function index()
+    {
+        $page = Param::get('page', 1);
+        $per_page = 10;
```
Please avoid using magic numbers po. You may declare it as constant. e.g., 

```
const CURRENT_PAGE = 10;
...
$per_page = self::CURRENT_PAGE;
```

Please apply to all po :bow:

rdaitan replied:
Done po.

maragatchalian commented on app/controllers/thread_controller.php:
```diff
@@ -0,0 +1,78 @@
+        notAuthRedirect('user/authenticate');
+
+        $thread  = new Thread;
+        $comment = new Comment;
```
Same as my comment here po: https://github.com/rdaitan/dietcake-board/pull/1/files#r38389442

rdaitan replied:
Done po.

maragatchalian commented on app/models/comment.php:
```diff
@@ -0,0 +1,55 @@
+    public static function countAll($thread_id)
+    {
+        $db = DB::conn();
+        return (int) $db->value("SELECT COUNT(*) FROM comment WHERE thread_id={$thread_id}");
```
You may remove `(int)` since the result of `COUNT` is always an integer :bow:

rdaitan replied:
Done po.

maragatchalian commented on app/models/comment.php:
```diff
@@ -0,0 +1,55 @@
+    public static function getAll($thread_id, $offset, $limit)
+    {
+        $db = DB::conn();
+        $rows = $db->rows("SELECT * FROM comment WHERE thread_id={$thread_id} ORDER BY created ASC LIMIT {$offset}, {$limit}");
```
You may remove `ASC` since the `ORDER BY` keyword in MySQL sorts the records in ascending order by default. :bow:

rdaitan replied:
done po.

maragatchalian commented on app/models/comment.php:
```diff
@@ -0,0 +1,55 @@
+
+        $comments = array();
+        foreach ($rows as $row) {
+            $comments[] = new Comment($row);
```
You may change this to:
`$comments[] = new self($row);`

since you're already in the Comment model po.
Please apply to all po. :bow:

rdaitan replied:
done po.

maragatchalian commented on app/models/thread.php:
```diff
@@ -0,0 +1,58 @@
+
+        $threads = array();
+        foreach ($rows as $row) {
+            $threads[] = new Thread($row);
```
Same as my comment here po: https://github.com/rdaitan/dietcake-board/pull/1/files#r38389981

rdaitan replied:
Done po.

maragatchalian commented on app/models/thread.php:
```diff
@@ -0,0 +1,58 @@
+    public static function countAll()
+    {
+        $db = DB::conn();
+        return (int) $db->value("SELECT COUNT(*) FROM thread");
```
Same as my comment here po: https://github.com/rdaitan/dietcake-board/pull/1/files#r38389798

rdaitan replied:
Done po.

darwinnavarro commented on app/app_exception.php:
```diff
@@ -1,4 +1,4 @@
-class AppException extends Exception
-{
-}
+class AppException extends Exception {}
```
Please follow this format po : 

``` php
class AppException extends Exception
{
}
```

rgmtan commented on app/controllers/user_controller.php:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    const ERROR_VAR                 = 'error';
+    const ERROR_MESSAGE_USERPASS    = 'Incorrect username or password';
```
please declare your variables at the top of your class

rdaitan replied:
Noted po. :bow:

RomeoVargas commented on app/config/sql/board2.sql:
```diff
@@ -0,0 +1,48 @@
+
+CREATE TABLE IF NOT EXISTS thread (
+    id        INT UNSIGNED NOT NULL AUTO_INCREMENT,
+    title     VARCHAR(255),
```
Does this mean that you're accepting titles up to 255 characters po?
If not, you could minimize the required size for title here po since you already have a validation for the title length :bow:

rdaitan replied:
No, sir. I'm not accepting titles up to 255 characters. I'll change it po.

RomeoVargas commented on app/helpers/html_helper.php:
```diff
@@ -5,3 +5,62 @@ function eh($string)
+
+function redirect($str) {
+    header("Location: " . url($str));
+    die();
```
i think this is unnecessary po :bow:

rdaitan replied:
Is it the die() po? I read that you should call it to prevent the script from executing any further. :bow:

RomeoVargas replied:
ahh, you mean if redirection via header() does not work you'll use die() po?
by using header() here you're already loading the controller of the url you entered po

rdaitan replied:
Not just when the header() fails po. I read that every time you set the location header, die() or exit() should be called to stop script execution. 
http://stackoverflow.com/questions/8665985/php-utilizing-exit-or-die-after-headerlocation

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,103 @@
+    {
+        $id;
+
+        if(array_key_exists(self::AUTH_USER_KEY, $_SESSION)) {
```
you could reverse this condition and make an early return instead of initializing $id po :bow:

rdaitan replied:
I never thought of that. Thank you sir. :bow:

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,103 @@
+
+        // Get the user by id
+        $row = $db->row('SELECT * FROM user WHERE username=?', array($username));
+        if(empty($row)) return false;
```
please practice using curly braces for if conditions even if they are one-liner so that it's more readable :bow: 

``` php
if (condition) {
    ...
}
```

rdaitan replied:
Noted po.

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,103 @@
+        );
+
+        $db = DB::conn();
+        $db->insert('user', $info);
```
you could directly create the array inside $db->insert here po, no need to create an extra variable

rdaitan replied:
I see po. Thank you.

rgmtan commented on app/controllers/thread_controller.php:
```diff
@@ -0,0 +1,79 @@
+                try {
+                    $thread->create($comment);
+                }
+                catch(ValidationException $e) {
```
catch should be on same line as closing brace of try

rdaitan replied:
Formatted na po. Thank you.

rgmtan commented on app/controllers/comment_controller.php:
```diff
@@ -0,0 +1,33 @@
+                }
+                break;
+            default:
+                throw new NotFoundException("{$page} is not found");
```
NotFoundException is not declared

rdaitan replied:
Declared na po.

RomeoVargas replied:
please be specific in your exceptions po :bow: 
what is NotFound here po?

rdaitan replied:
PageNotFoundException po.

rgmtan commented on app/models/thread.php:
```diff
@@ -0,0 +1,59 @@
+        $db = DB::conn();
+
+        $row = $db->row('SELECT * FROM thread WHERE id=?', array($id));
+        if(empty($row)){
```
``` php
if (!$row) {
}
```

RomeoVargas commented on app/helpers/validation_helper.php:
```diff
@@ -0,0 +1,21 @@
+
+function validate_unique_name($username)
+{
+    return !User::getByUsername($username);
```
instead of creating a validation helper for unique username, why not make username in user table unique then just catch the SimpleDBI Exception for it po?

rdaitan replied:
If I just catch the exception, the username's uniqueness will not be evaluated with the other fields. So if the other fields did not pass the evaluation, the username's uniqueness will not be evaluated and an error will not be shown even if it already exists po.

![screen shot 2015-09-02 at 13 52 45](https://cloud.githubusercontent.com/assets/13476640/9623922/2dae169c-517a-11e5-9432-cac86416e8fc.png)

RomeoVargas replied:
I see. You want to include this to the validation error so that it's easier in the View part po?

rdaitan replied:
Yes, sir.

RomeoVargas replied:
ahh if that's the case its ok po :bow:

rdaitan replied:
Thank you po. :bow:

fatimaPHP replied:
I think that username/email should be defined unique in the table.
It will not affect the validation since you call `validate()` first before inserting to the table.
Please check the example below on how to check pdo exceptions when there's duplicate entry

``` php
# user.php
if ($this->validate()) {
    throw new ValidationException('Invalid input!');
}

$params = array(
    ...
);

try {
    $db->insert('user', $params);
} catch (PDOException $e) {
    // duplicate entry, do something else
    if ($e->errorInfo[1] == 1062) {
        throw new DuplicateEntryException;
    }
}
```

rdaitan replied:
How about this po?
16b423459f9ad64926e7ffa8db94eef3ebe2ee79

fatimaPHP replied:
The transaction is unnecessary po since you are only writing to one table. 
Besides the order of process is weird po, normally you validate input and if it is correct you then insert it in the database. Without the transaction, you cannot rollback if there is an invalid input.

fatimaPHP replied:
Maybe, you can just catch the duplicate entry exception from the controller and set the $user property validation_errors to true so you can display the error message in the view po :bow:

``` php
try {
...
} catch (DuplicateEntryException $e) {
    $user->validation_errors['username']['unique'] = true;
}
```

rdaitan replied:
How about this?
4486b2d3f8b22776c2d47af46c5cf9102c32efd1

fatimaPHP replied:
:+1:

rdaitan replied:
Thank you. :bow:

RomeoVargas commented on app/models/comment.php:
```diff
@@ -0,0 +1,55 @@
+    public static function countAll($thread_id)
+    {
+        $db = DB::conn();
+        return $db->value("SELECT COUNT(*) FROM comment WHERE thread_id={$thread_id}");
```
please use placeholders :bow: 

``` php
$db->value('SELECT COUNT(*) FROM comment WHERE thread_id = ?', array($thread_id));
```

RomeoVargas replied:
apply this to all cases po :bow:

rdaitan replied:
Done po.

RomeoVargas commented on app/models/thread.php:
```diff
@@ -0,0 +1,59 @@
+    {
+        $this->validate(); // before calling this method, $this->title was set first
+        $comment->validate();
+        if($this->hasError() || $comment->hasError()) {
```
you could directly call validate() for thread and comment here po :bow: 

``` php
if (!$this->validate() || !$comment->validate()) {
    ...
}
```

rdaitan replied:
But if I do this, $comment->validate() will not execute if !$this->validate() evaluates to true. If this happens, no error message will appear regarding the comment.

![selection_461](https://cloud.githubusercontent.com/assets/13476640/9621977/2ccbc126-5160-11e5-9069-beab3bd39f07.jpeg)

I used bitwise OR to retain the functionality po. :bow:

``` php
if(!$this->validate() | !$comment->validate()) {
    // before calling $this->validate(), $this->title was set first
    throw new ValidationException('Invalid thread or comment.');
}
```

RomeoVargas replied:
Good job po :+1:

rdaitan replied:
Thank you, sir. :smile:

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,96 @@
+
+        // Get the user by username
+        $row = $db->row('SELECT * FROM user WHERE username=?', array($username));
+        if(empty($row)) {
```
please refrain from using empty($variable), use !$variable instead po :bow: 
also for conditional statements, be mindful of spacing

``` php
if (condition) {
    ...
}
```

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,96 @@
+        $hashedPassword = bhash($password, $salt);
+        $passwordsMatch = $hashedPassword === $user->password;
+
+        if($passwordsMatch) {
```
there's no need for additional variable for passwordsMatch here po :bow: 
also this could be an early return as well

``` php
if ($hashedPassword !== $user->password) {
    return false;
}

$user->setAuthenticated();
return true;
```

rdaitan replied:
I see po.

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,96 @@
+    {
+        if(!array_key_exists(self::AUTH_USER_KEY, $_SESSION)) {
+            return false;
+        } else {
```
no need for an else { } here po since you're already returning false in your if statement

rdaitan replied:
Okay po.

RomeoVargas commented on app/controllers/user_controller.php:
```diff
@@ -0,0 +1,74 @@
+        if(User::getAuthenticated()) {
+            session_destroy();
+        }
+        redirect('/');
```
maybe it's better to input the proper url here so that it's easier to know where will this be redirected po :bow:

rdaitan replied:
Done po. :bow:

RomeoVargas commented on app/controllers/thread_controller.php:
```diff
@@ -0,0 +1,78 @@
+        $page    = Param::get('page_next', 'create');
+
+        switch($page) {
+            case 'create':
```
``` php
switch ($page) {
case 'create':
    ...
    break;
case 'create_end':
    ...
    break;
default:
    ...
    break;
}
```

RomeoVargas replied:
please apply to all switch-case :bow:

rdaitan replied:
Noted po. :bow:

rdaitan replied:
Done po. :bow:

RomeoVargas commented on app/models/comment.php:
```diff
@@ -0,0 +1,55 @@
+    public static function getAll($thread_id, $offset, $limit)
+    {
+        $db = DB::conn();
+        $rows = $db->rows("SELECT * FROM comment WHERE thread_id=? LIMIT {$offset}, {$limit}", array($thread_id));
```
Since you're already using placeholder for thread_id, why not use placeholders as well for the offset and limit?

rdaitan replied:
I tried po. But the values of $offset and $limit are being wrapped in single quotes, making the query:

``` sql
SELECT * FROM comment WHERE thread_id=? LIMIT '0', '10'
```

Nag-eerror po.

RomeoVargas replied:
ahh i see po, you can use sprintf() instead po 

``` php
sprintf("SELECT * FROM comment WHERE thread_id = %d LIMIT %d, %d", $thread_id, $offset, $limit)
```

rdaitan replied:
I see. Thank you po.
:bow:

RomeoVargas replied:
or you could try this out po and see if this works

``` php
$db->search('comment', 'thread_id = ?', array($thread_id), null, array($offset, $limit));
```

rdaitan replied:
It worked. I thought it was deprecated because I saw '@DEPRECATED' when I looked in the SimpleDBI.php so I decided not to use it at first.

RomeoVargas commented on app/models/thread.php:
```diff
@@ -0,0 +1,58 @@
+        $this->id = $db->lastInsertId();
+        $comment->create($this);
+
+        $db->commit();
```
I can see that you've used a transaction here po, you should enclose this in a try .. catch block so that if there's a problem in insertion of record, you could rollback the transaction

rdaitan replied:
I see po. I'll do that. :bow:

rdaitan committed:
Enclosed thread transaction in try...catch

> I can see that you've used a transaction here po, you should
  enclose this in a try .. catch block so that if there's a
  problem in insertion of record, you could rollback the
  transaction

RomeoVargas commented on app/controllers/comment_controller.php:
```diff
@@ -0,0 +1,32 @@
+            $comment->body      = Param::get('body');
+            $comment->user_id   = User::getAuthenticated()->id;
+
+            try{
```
Please double check all spacings po :bow: 

``` php
try {
    ...
} catch (ValidationException $e) {
    ...
}
```

rdaitan committed:
Fixed spacing

rdaitan committed:
Fixed spacing

rdaitan committed:
Spacing

rdaitan committed:
Revert "Spacing"

This reverts commit 3270b3b1877ac1d5e0cd19f1c27f9251cafe701b.

rdaitan committed:
Used sprintf instead of variable expansion

RomeoVargas commented on app/models/user.php:
```diff
@@ -0,0 +1,96 @@
+
+        $hashedPassword = bhash($password, $salt);
+
+        if(!$hashedPassword === $user->password) {
```
i think this is easier to read po :bow: 

``` php
if ($hashedPassword !== $user->password) {
```

also please check the spacings po

``` php
if (condition) {
```

rdaitan replied:
I'll change it po.

rdaitan committed:
Fixed authentication condition

What I did is even wrong.

rdaitan committed:
Fixed spacing.

rdaitan committed:
Fixed spacing.

rdaitan committed:
Merge commit '37c1704' into dev_1.0-code_review

rdaitan committed:
Spacing in switch case

rdaitan committed:
Used User::getByUsername() on authentication

Old code was replaced by getByUsername to get the user.

rdaitan committed:
Removed variable on User::getAuthenticated()

No need for a variable because only the return statement will use
it. Besides, it was pretty obvious that it is the user id that is
being retrieved from the $_SESSION array.

rdaitan committed:
Renamed constant AUTH_USER_KEY

User::AUTH_USER_KEY seems redundant. User::AUTH_SESS_KEY seems to
reflect what it is better--a key in the $_SESSION array.

rdaitan committed:
Spacing and arrangement

rdaitan committed:
Static functions placed on top

rdaitan committed:
Static functions on top. Removed comment.

Only in user.php.

rdaitan committed:
Fixed spacing and alignment

rdaitan committed:
Revert "Fixed spacing."

This reverts commit 331079a462f2855c21129a033a05c0e0d09aea3d.

Conflicts:

	app/models/comment.php

rdaitan committed:
Broke very long lines of codes

rdaitan committed:
Fixed spacing

rdaitan committed:
Improved readability

rdaitan committed:
Defined new constant, CRYPT_BFISH_SALT_LENGTH

Is using a magic number in User::authenticate(). Defined new constant
to fix that.

rdaitan committed:
Used new constant

Line is too long as well, so I broke it down.

rdaitan committed:
Improved readability

Improved readability.

Spacing

rdaitan committed:
Renamed redirect helpers

Changed name of authentication redirect helpers

Names are a bit of a mouthful for now, but they are more descriptive.

Renamed redirect helpers.

Thanks, Max and Renzo.

rdaitan committed:
Renamed pagination helper

maragatchalian commented on app/controllers/user_controller.php:
```diff
@@ -0,0 +1,74 @@
+    {
+        authRedirect();
+
+        $user = new User;
```
`$user = new User();`

rdaitan replied:
Done po.

rdaitan committed:
Followed standard

maragatchalian commented on app/models/comment.php:
```diff
@@ -0,0 +1,57 @@
+class Comment extends AppModel
+{
+    public $validation = array(
+        'body' => array('length' => array('validate_between', 1, 200)),
```
Please eliminate all magic numbers po. Use may use constants.

```
MIN_BODY_LENGTH = 1;
MAX_BODY_LENGTH = 200;
...
        'body' => array('length' => array('validate_between', self::MIN_BODY_LENGTH, self::MAX_BODY_LENGTH)),
...
```

rdaitan replied:
Okay po.

rdaitan committed:
Removed magic numbers in models

rdaitan committed:
Formatting

rdaitan committed:
Code formatting

rdaitan committed:
Formatting changes

rdaitan committed:
Formatting

rdaitan committed:
Redefined bfish constants

Names are a bit too long.

rdaitan committed:
Formatted code
Code Formatting

Code Formatting

Code Formatting

Code Formatting

fatimaPHP commented on app/models/comment.php:
```diff
@@ -0,0 +1,61 @@
+    {
+        $db     = DB::conn();
+        $rows   = $db->rows(
+            sprintf("SELECT * FROM comment WHERE thread_id=%d LIMIT %d, %d", $thread_id, $offset, $limit)
```
This would be vulnerable to sql injection because thread_id was not passed as parameter to the prepared statement :bow:

rdaitan replied:
I'll fix it right now. What about $offset and $limit? I also tried passing them as parameters before, but they only cause an error in the statement. Thank you. :bow:

fatimaPHP replied:
`sprintf` is ok po since you cannot use placeholder(?) after the LIMIT clause

fatimaPHP commented on app/controllers/comment_controller.php:
```diff
@@ -0,0 +1,32 @@
+        case 'create_end':
+            $comment            = new Comment();
+            $comment->body      = Param::get('body');
+            $comment->user_id   = User::getAuthenticated()->id;
```
If getAuthenticated returns false, it will cause an error since id is an undefined property

rdaitan replied:
Noted, ma'am.

fatimaPHP commented on app/controllers/user_controller.php:
```diff
@@ -0,0 +1,74 @@
+            } else {
+                $this->set(array(self::ERROR_VAR => self::ERROR_MESSAGE_USERPASS));
+                $page = 'auth';
+            }
```
you may remove L52-53 from the else block po

rdaitan replied:
Done po. :bow:

fatimaPHP commented on app/helpers/html_helper.php:
```diff
@@ -2,6 +2,81 @@
+    return $str;
+}
+
+function redirect($str)
```
how about rename $str to $url po ?

rdaitan replied:
Done po. :bow:

rdaitan replied:
Okay po. :bow:

fatimaPHP commented on app/views/thread/index.php:
```diff
@@ -0,0 +1,40 @@
+                <h3>Welcome, <?php eh($auth_user->username); ?>!</h3>
+            <?php } else { ?>
+                <h3>Hello, guest!</h3>
+            <?php } ?>
```
This looks weird po, please use the shorthand version po

``` php
<?php if (condition): ?>
    # code block
<?php else: ?>
    # code block
<?php endif; ?>
```

rdaitan replied:
Done po.

fatimaPHP commented on app/views/thread/view.php:
```diff
@@ -0,0 +1,48 @@
+            <a href='<?php eh(url(APP_URL)); ?>'>&larr; All threads</a>
+        </div>
+        <div class="offset-top">
+            <?php foreach ($comments as $n => $comment) {?>
```
please use the shorthand for this

``` php
<?php foreach(...): ?>
<?php endforeach; ?>
```

rdaitan replied:
Done. :bow:

rdaitan committed:
Passed $thread_id as parameter in query

> This would be vulnerable to sql injection because thread_id was
  not passed as parameter to the prepared statement

rdaitan committed:
Performed check if result of User::getAuthenticated is false

> If getAuthenticated returns false, it will cause an error since
  id is an undefined property

rdaitan committed:
Removed lines 52 and 53 outside of else block

> you may remove L52-53 from the else block po

rdaitan committed:
Renamed parameters

> how about rename $str to $url po ?

rdaitan committed:
Used shorthand of if and foreach

rdaitan committed:
discussion_r38842523 fix 3

If a ValidationException is thrown before attempting to insert into
the database, the user will not know if the username has a duplicate.

rdaitan committed:
Defined constants for URLs

Avoiding magic strings.
